Services/WindowServer/Compositor.cpp

```diff
@@ -111,21 +111,23 @@ void Compositor::compose()
         m_wallpaper_mode = mode_to_enum(wm.config()->read_entry("Background", "Mode", "simple"));
     auto& ws = Screen::the();
 
-    auto dirty_rects = move(m_dirty_rects);
-
-    if (dirty_rects.size() == 0) {
+    if (!m_invalidated_any) {
         // nothing dirtied since the last compose pass.
         return;
     }
 
-    dirty_rects.add(Gfx::IntRect::intersection(m_last_geometry_label_rect, Screen::the().rect()));
-    dirty_rects.add(Gfx::IntRect::intersection(m_last_cursor_rect, Screen::the().rect()));
-    dirty_rects.add(Gfx::IntRect::intersection(m_last_dnd_rect, Screen::the().rect()));
-    dirty_rects.add(Gfx::IntRect::intersection(current_cursor_rect(), Screen::the().rect()));
+    auto dirty_screen_rects = move(m_dirty_screen_rects);
+    dirty_screen_rects.add(Gfx::IntRect::intersection(m_last_geometry_label_rect, Screen::the().rect()));
+    dirty_screen_rects.add(Gfx::IntRect::intersection(m_last_dnd_rect, Screen::the().rect()));
+    if (m_invalidated_cursor) {
+        // TODO: what is this?
```

Comment:
Drag n drop!

Reply:
Ah, thanks! I kinda wish all these special things (desktop, dnd, ...) were just windows themselves, then we wouldn't need these special cases and the overhead that comes with it. And looks like they're not quite working right, I'm guessing I just need to add the previous position to the `flush_rects`. I'll look at it later tonight.

Reply:
Agreed that it would be sweet if they were just windows.